include/net/mqtt_socket.h

```diff
@@ -0,0 +1,655 @@
+	mqtt_suback_param_t suback;
+
+	/** Parameters accompanying MQTT_EVT_UNSUBACK event. */
+	mqtt_unsuback_param_t unsuback;
+} mqtt_evt_param_t;
+
+/**@brief Defined MQTT asynchronous event notified to the application. */
+typedef struct {
+	/** Identifies the event. */
+	mqtt_evt_id_t id;
+
+	/** Contains parameters (if any) accompanying the event. */
+	mqtt_evt_param_t param;
+
+	/** Event result. For example, MQTT_EVT_CONNACK has a result code
+	 *  indicating success or failure code of connection procedure.
+	 */
+	int result;
+} mqtt_evt_t;
+
+/**@brief Asynchronous event notification callback registered by the application
+ *        with the module to receive module events.
+ *
+ * @param[in] client Identifies the client for which the event is notified.
+ * @param[in] evt Event description along with result and associated
+ *                parameters (if any).
+ */
+typedef void (*mqtt_evt_cb_t)(mqtt_client_t *const client,
+			      const mqtt_evt_t *evt);
+
+/**@brief MQTT transport type. */
+typedef enum {
+	/** Use non secure TCP transport for MQTT connection. */
+	MQTT_TRANSPORT_NON_SECURE       = 0x00,
+
+	/** Use secure TCP transport (TLS) for MQTT connection. */
+	MQTT_TRANSPORT_SECURE           = 0x01,
+
+	/** Shall not be used as a transport type.
+	 *  Indicator of maximum transport types possible.
+	 */
+	MQTT_TRANSPORT_MAX              = 0x02
+} mqtt_transport_type_t;
+
+/**@brief MQTT transport specific data.
+ */
+typedef struct {
+	/** Transport type selection for client instance.
+	 *  @ref mqtt_transport_type_t for possible values. MQTT_TRANSPORT_MAX
+	 *  is not a valid type.
+	 */
+	mqtt_transport_type_t type;
+
+	union {
+		/* TCP socket transport for MQTT */
+		struct {
+			/** Socket descriptor. */
+			int sock;
+		} tcp;
+	};
+} mqtt_transport_t;
+
+/**@brief MQTT Client definition to maintain information relevant to the client.
+ */
+struct mqtt_client_t {
+	/** MQTT transport configuration and data. */
+	mqtt_transport_t transport;
+
+	/** Unique client identification to be used for the connection.
+	 *  Shall be zero length or NULL valued.
+	 */
+	mqtt_client_id_t client_id;
+
+	/** Broker details, for example, address, port. Address type should
+	 *  be compatible with transport used.
+	 */
+	const void *broker;
+
+	/** User name (if any) to be used for the connection. NULL indicates
+	 *  no user name.
+	 */
+	mqtt_username_t *user_name;
+
+	/** Password (if any) to be used for the connection. Note that if
+	 *  password is provided, user name shall also be provided. NULL
+	 *  indicates no password.
+	 */
+	mqtt_password_t *password;
+
+	/** Will topic and QoS. Can be NULL. */
+	mqtt_topic_t *will_topic;
+
+	/** Will message. Can be NULL. Non NULL value valid only if will topic
+	 *  is not NULL.
+	 */
+	mqtt_will_message_t *will_message;
+
+	/** Application callback registered with the module to get MQTT events.
+	 */
+	mqtt_evt_cb_t evt_cb;
+
+	/** MQTT protocol version. */
```

Comment:
It looks like there would be holes in the struct, what about ordering the fields so that the u8_t and the bitfields are at the end.